### PR TITLE
RFC: Audit CNAMEs

### PR DESCRIPTION
**Pull request by kontrolldkaos**
Moving over and following new instructions for RFC

### CODE REVIEW

**Comment by pauljamescleary** [comment]
I like this idea.  I would like to suggest adding some ideas around the implementation bits.  My comments are:

1. I imagine it would make sense to put a marker on the record in vinyldns to indicate that it cannot resolve.  This marker should be propagated through the API as well as the Portal
1. In the portal, we can flag the CNAMEs like we do with dotted hosts (warning).  We should provide somewhere a description of the issue.
1. We will have a new API endpoint to "audit" a zone.  I imagine in the future we may have different "audit" types, CNAME being one of them, mis-matching PTRs being another one of them.  We could default to `all` if no audit type was provided.
1. We should add the ability to run audits in the Portal.  Perhaps a new tab called `Audit` on the zone.  There would be only a single audit which would be `CNAME Resolution`.

**Issues**
- From a UI perspective, an AUDIT could take a long time to complete so it would be background / async.  In that case, we would need to _save_ the audit.
- We have to actively prune audits so they don't get too big, so perhaps we limit to last `N` to a sane number, say last `10` reports?

**Comment by kontrolldkaos** [comment]
### Issues / Questions

**1. From a UI perspective, an AUDIT could take a long time to complete so it would be background / async. In that case, we would need to save the audit.**
   
Yes I agree that's the way to go so there is not a long wait for the end user. 


**2. We have to actively prune audits so they don't get too big, so perhaps we limit to last N to a sane number, say last 10 reports?**

I think that this could be configurable, but possibly no more than 5 - 10 per zone. If there is a need to save more audits, there should be an option to export, save offline, or emailed. 

**3. Should we interrogate the DNS Master for this (which could lead to a lot of queries), rely on plain old DNS lookup from the VinylDNS host, or perhaps make this configurable?**

This is another option that could be configurable. Some people would like to know the different responses from internal DNS servers vs external and public DNS servers.

**Comment by kontrolldkaos** [comment]
I also would like to add that I agree with the UI approach where it's similar to the dotted hosts. It could be ran in-line and indicate that the record has an issue.

File: text/0000-audit-cnames.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+database.
+
+2. Do DNS checks on CNAME data gathered and look for NXDOMAIN, No Answers, and Timeouts, then 
+report.
+
+3. Use back end data generated to report to VinylDNS UI for the end user.
+
+4. Use back end data generated to generate reports on set intervals. 
+
+
```
Comment (pauljamescleary) [inline text/0000-audit-cnames.md]: Want to discuss some design in here (this is more of a technical discussion than a requirement discussion).  @kontrolldkaos may submit a PR to this RFC, stay tuned.

The process would look like...

1. Provide a way for users to request a CNAME audit.  Something like `POST /zones/<id>/audit?type=CNAME`
1. When the audit is received, validate zone access and place on the queue.
1. Modify the existing backend processor to handle a "CNAME AUDIT" message type
1.  ? we need to flush out the workflow here ?
1. Once the audit is complete, save the results in the database

Some requirements to discuss:

1. Need a new `ZoneAuditRepository` that allows us to save audit results
1. Need a new `ZoneCommand` and protocol buffer so we can send the message
1. Need to add a new `ZoneService.audit` function, that takes an `Audit` as a parameter.
1. Need to add a new `Audit` tab to the portal.  The tab should list past audits, and provide a button to allow the user to start a new audit
1. We should limit how often each audit type can be performed (like we do with zone syncing)

**Comment by pauljamescleary** [comment]
Proposing we postpone this for the time being as our short-term roadmap is around batch changes and zone management.

We also have to explore the idea of "Admin" users further to support this kind of feature

**Comment by kontrolldkaos** [comment]
I agree. We can discuss this in one of our upcoming meetings.

**Comment by pauljamescleary** [comment]
Closing this with a disposition of postpone

**Pull request closed without merging**